draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+~~~~~~~~~~~~~~~
+
+I/O bound tasks are tasks which mainly perform a network operation or
+a disk operation.
+
+I/O bound tasks are specifically marked as such using Python's `async def`
+notation for defining awaitable functions. They will run in a Python coroutine.
+
+Due to that, any I/O operation in that task must be asynchronous in order to
+avoid blocking the event loop.
+
+Some of the user's asynchronous tasks won't use trio as their event loop
+but will use the more commonly used asyncio event loop which we do support.
+
+In that case, the user must specify the event loop they are going to use for the
+task.
+
+CPU Bound Tasks
+~~~~~~~~~~~~~~~
+
+CPU bound tasks are tasks which mainly perform a calculation of some sort such
+as calculating an average, hashing, serialization or deserialization,
+compression or decompression, encryption or decryption etc.
+In some cases where no asynchronous code for the I/O operation is available
+CPU bound tasks are also an appropriate choice as they will not block
+the event loop for the duration of the task.
+
+Performing operations which release the :term:`GIL` is recommended to avoid
+throttling the concurrency of the worker.
+
+CPU bound tasks are specifically marked as such using Python's
+`def` notation for defining functions. They will run in a Python thread.
+
+Using threads instead of forking the main process has its upsides:
+
+- It simplifies the Worker's architecture and makes it less brittle.
+
+  Processes require :term:`IPC` to communicate with each other.
+  This complicates implementation since multiple methods are required to support
+  :term:`IPC` reliably across all operating systems Celery supports.
+  Threads on the other hand require less complicated means of communication.
+
+  In `trio`_, we simply use a memory channel which is a coroutine and thread safe
+  way to send and receive values.
+
+- PyPy's JIT warms up faster.
+
+  When using PyPy, using threads means that we get to keep our previous JIT traces
+  and therefore JIT warmup will occur faster.
+
+  If we'd use processes, each process has to warm up its own JIT which results
+  in tasks being executed slower for a longer period of time.
+
+Using threads for CPU bound tasks unfortunately has some downsides as well:
+
+- Pure Python CPU bound workloads cannot be executed in parallel.
+
+  In both CPython and PyPy the :term:`GIL` prevents executing two Python bytecodes
+  in parallel by design.
+
+  This results in slower execution of Python code when using threads.
+
+- The :term:`GIL`'s implementation in CPython 3.x has a defect in design.
+
+  According to a `bug report <https://bugs.python.org/issue7946>`_ the new GIL
+  in Python 3 CPU bound threads may starve I/O threads (in our case the main thread).
+
+  .. note ::
+
+    This is not an issue with PyPy's implementation of the `GIL`_
+    `according to Armin Rigo <https://bugs.python.org/msg346495>`_, PyPy's creator.
+
+  - Tasks are no longer isolated.
```

Comment:
This is quite important, I am not sure we can overcome this limitation. I think sub-processes should remain an option. An additional argument is that memory is not easily released on the same interpreter instance and this is mitigated by forking new sub-processes in regular intervals or after a specific amount of tasks. I believe this is also the case with most web application servers (such as uWSGI).

Reply:
Boxed Tasks are the way to isolate execution of a task.
The Boxed Workers are not forked from the main Worker process but are started separately. 
I haven't really described them yet but essentially they are Workers with a concurrency budget of 1.